system/jlib/jexcept.cpp

```diff
@@ -991,6 +991,7 @@ void excsighandler(int signum, siginfo_t *info, void *extra)
     signal(SIGBUS, SIG_DFL);
     signal(SIGILL, SIG_DFL);
     signal(SIGFPE, SIG_DFL);
+    signal(SIGABRT, SIG_DFL);
```

Comment:
If this signal isn't added then the SIGABRT flows through HPCC System without any footprint. In simplified view a workunit can disappear in the thin air. 
If you run this short code:
```
    boolean seg() := beginc++ #option action
       #include <csignal>
       #include <csignal>
       #body
       raise(SIGABRT);
       return false;
   endc++;
   output(seg());
```
as mycrash.ecl with the ecl command, then the output is:
``` 
ecl run -v -t hthor mycrash.ecl
EXEC: Creating PIPE program process : 'eclcc -E "mycrash.ecl"' - hasinput=0, hasoutput=1 stderrbufsize=0
EXEC: Pipe: process 16546 complete 0

Deploying ECL Archive mycrash.ecl

Deployed
   wuid: W20160121-192321
   state: compiled

Running deployed workunit W20160121-192321
W20160121-192321 failed
<Result>
 <Exception><Code>1000</Code><Source></Source><Message>Workunit terminated unexpectedly</Message></Exception>
</Result>
```
So, you have no idea what happened. Let's see the logs. Relevant part from eclagent log is:
```
00000000 2016-01-21 19:23:22.874 16599 16599 "Logging to /var/log/HPCCSystems/myeclagent/eclagent.2016_01_21.log"
00000001 2016-01-21 19:23:22.886 16599 16599 "ECLAGENT build community_6.1.0-trunk0Debug[heads/HPCC-14749-improvement-0-ga76bd0-dirty]"
00000002 2016-01-21 19:23:22.897 16599 16599 "Setting workunit trace level to 10"
00000003 2016-01-21 19:23:22.897 16599 16599 "openWorkUnit W20160121-192321"
00000004 2016-01-21 19:23:22.903 16599 16599 "Entering doProcess ()"
00000005 2016-01-21 19:23:22.903 16599 16599 "Waiting for workunit lock"
00000006 2016-01-21 19:23:22.903 16599 16599 "Obtained workunit lock"
00000007 2016-01-21 19:23:22.913 16599 16599 "Loading dll (libW20160121-192321.so) from location /var/lib/HPCCSystems/myeclccserver/libW20160121-192321.so"
00000008 2016-01-21 19:23:22.914 16599 16599 "Starting process"
00000009 2016-01-21 19:23:22.914 16599 16599 "RoxieMemMgr: Setting memory limit to 4294967296 bytes (16384 pages)"
0000000A 2016-01-21 19:23:22.914 16599 16599 "Transparent huge pages used for roxiemem heap"
0000000B 2016-01-21 19:23:22.914 16599 16599 "Memory released to OS in 8192k blocks"
0000000C 2016-01-21 19:23:22.915 16599 16599 "RoxieMemMgr: 16384 Pages successfully allocated for the pool - memsize=4294967296 base=0x7f7befe00000 alignment=262144 bitmapSize=512"
0000000D 2016-01-21 19:23:22.915 16599 16599 "Waiting for run lock"
0000000E 2016-01-21 19:23:22.916 16599 16599 "Obtained run lock"
0000000C 2016-01-21 19:35:04.810 15184 15185 "Exiting agentexec"
```
That's it. No more info about why. The ESP server log more shorter:
```
000000C2 2016-01-21 19:23:22.823 15622 16597 "Agent request 'W20160121-192321' enqueued on 'hthor.agent'"
000000C3 2016-01-21 19:23:42.911 15622 16597 "WARNING: checkAbnormalTermination: workunit terminated: 3100966396977 state = 4"
000000C4 2016-01-21 19:23:42.923 15622 16597 "Time taken for CSoapService::processRequest(): 40161477184 cycles (40161M) = 20113 msec"
```

Yes, you are right SIGABRT generates core anyway, but it comes from ECL code and we have absolute no idea what happened because we have no any info in our logs. 

Your question is why I would like to add SIGABRT to our signal handling? My answer is this: (execute same ECL code as above, but we are handling the SIGABRT signal):

The output of the ecl command now gives some information about what happened:
```
ecl run -v -t hthor mycrash.ecl
EXEC: Creating PIPE program process : 'eclcc -E "mycrash.ecl"' - hasinput=0, hasoutput=1 stderrbufsize=0
EXEC: Pipe: process 3644 complete 0

Deploying ECL Archive mycrash.ecl

Deployed
   wuid: W20160121-193910
   state: compiled

Running deployed workunit W20160121-193910
W20160121-193910 failed
<Result>
 <Exception><Code>1000</Code><Source>ECLAGENT</Source><Message>SIG: Aborted(6), accessing 000003E600000E73, IP=00007F1C04D65CC9</Message></Exception>
</Result>
```
and the eclagent log contains much more information:
```
00000000 2016-01-21 19:39:11.262  3699  3699 "Logging to /var/log/HPCCSystems/myeclagent/eclagent.2016_01_21.log"
00000001 2016-01-21 19:39:11.276  3699  3699 "ECLAGENT build community_6.1.0-trunk0Debug[heads/HPCC-14749-improvement-0-ga76bd0-dirty]"
00000002 2016-01-21 19:39:11.295  3699  3699 "Setting workunit trace level to 10"
00000003 2016-01-21 19:39:11.296  3699  3699 "openWorkUnit W20160121-193910"
00000004 2016-01-21 19:39:11.302  3699  3699 "Entering doProcess ()"
00000005 2016-01-21 19:39:11.303  3699  3699 "Waiting for workunit lock"
00000006 2016-01-21 19:39:11.303  3699  3699 "Obtained workunit lock"
00000007 2016-01-21 19:39:11.314  3699  3699 "Loading dll (libW20160121-193910.so) from location /var/lib/HPCCSystems/myeclccserver/libW20160121-193910.so"
00000008 2016-01-21 19:39:11.315  3699  3699 "Starting process"
00000009 2016-01-21 19:39:11.315  3699  3699 "RoxieMemMgr: Setting memory limit to 4294967296 bytes (16384 pages)"
0000000A 2016-01-21 19:39:11.315  3699  3699 "Transparent huge pages used for roxiemem heap"
0000000B 2016-01-21 19:39:11.315  3699  3699 "Memory released to OS in 8192k blocks"
0000000C 2016-01-21 19:39:11.315  3699  3699 "RoxieMemMgr: 16384 Pages successfully allocated for the pool - memsize=4294967296 base=0x7f1aebe00000 alignment=262144 bitmapSize=512"
0000000D 2016-01-21 19:39:11.315  3699  3699 "Waiting for run lock"
0000000E 2016-01-21 19:39:11.317  3699  3699 "Obtained run lock"
0000000F 2016-01-21 19:39:11.318  3699  3699 "================================================"
00000010 2016-01-21 19:39:11.318  3699  3699 "Signal:    6 Aborted"
00000011 2016-01-21 19:39:11.318  3699  3699 "Fault IP:  00007F1C04D65CC9"
00000012 2016-01-21 19:39:11.318  3699  3699 "Accessing: 000003E600000E73"
00000013 2016-01-21 19:39:11.318  3699  3699 "Registers:"
00000014 2016-01-21 19:39:11.318  3699  3699 "EAX:0000000000000000  EBX:00007FFFA4075A30  ECX:FFFFFFFFFFFFFFFF  EDX:0000000000000006  ESI:0000000000000E73  EDI:0000000000000E73"
00000015 2016-01-21 19:39:11.318  3699  3699 "CS:EIP:0033:00007F1C04D65CC9"
00000016 2016-01-21 19:39:11.318  3699  3699 "   ESP:00007FFFA4075468  EBP:00007FFFA4075480"
00000017 2016-01-21 19:39:11.318  3699  3699 "Stack[00007FFFA4075468]: 00007F1BFC24528A A40754A000007F1B 00007FFFA40754A0 A4075A3800007FFF 00007FFFA4075A38 A40754C000007FFF 00007FFFA40754C0 FC2450B200007FFF"
00000018 2016-01-21 19:39:11.318  3699  3699 "Stack[00007FFFA4075488]: 00007F1BFC2450B2 013625A000007F1B 00000000013625A0 013625A000000000 00000001013625A0 A4075A3000000001 00007FFFA4075A30 0135FC6000007FFF"
00000019 2016-01-21 19:39:11.318  3699  3699 "Stack[00007FFFA40754A8]: 000000000135FC60 013625A000000000 00000001013625A0 A4075A3800000001 00007FFFA4075A38 A40754E000007FFF 00007FFFA40754E0 038029FC00007FFF"
0000001A 2016-01-21 19:39:11.318  3699  3699 "Stack[00007FFFA40754C8]: 00007F1C038029FC 0000000100007F1C 0000000100000001 01361F0000000001 0000000001361F00 A407559000000000 00007FFFA4075590 03801F8000007FFF"
0000001B 2016-01-21 19:39:11.318  3699  3699 "Stack[00007FFFA40754E8]: 00007F1C03801F80 0136B9C000007F1C 000000000136B9C0 0135645000000000 0000000101356450 0136232800000001 0000000001362328 01361F0000000000"
0000001C 2016-01-21 19:39:11.319  3699  3699 "Stack[00007FFFA4075508]: 0000000001361F00 0000000500000000 0000000000000005 0000000000000000 0000000000000000 A407554000000000 00007FFFA4075540 0371E28000007FFF"
0000001D 2016-01-21 19:39:11.319  3699  3699 "Stack[00007FFFA4075528]: 00007F1C0371E280 A4075A3000007F1C 00007FFFA4075A30 A40755E000007FFF 00007FFFA40755E0 02E78BBB00007FFF 00007F1C02E78BBB 05F1E51500007F1C"
0000001E 2016-01-21 19:39:11.319  3699  3699 "Stack[00007FFFA4075548]: 00007F1C05F1E515 01361F0000007F1C 0000000001361F00 F14A940000000000 4C88F50BF14A9400 000000014C88F50B 0000000000000001 A4075A3000000000"
0000001F 2016-01-21 19:39:11.319  3699  3699 "Backtrace:"
00000020 2016-01-21 19:39:11.320  3699  3699 "  /opt/HPCCSystems/lib/libjlib.so(_Z16printStackReportv+0x43) [0x7f1c0575605a]"
00000021 2016-01-21 19:39:11.320  3699  3699 "  /opt/HPCCSystems/lib/libjlib.so(_Z13excsighandleriP9siginfo_tPv+0x35e) [0x7f1c057559bd]"
00000022 2016-01-21 19:39:11.320  3699  3699 "  /lib/x86_64-linux-gnu/libc.so.6(+0x36d40) [0x7f1c04d65d40]"
00000023 2016-01-21 19:39:11.320  3699  3699 "  /lib/x86_64-linux-gnu/libc.so.6(gsignal+0x39) [0x7f1c04d65cc9]"
00000024 2016-01-21 19:39:11.320  3699  3699 "  /var/lib/HPCCSystems/myeclccserver/libW20160121-193910.so(+0x128a) [0x7f1bfc24528a]"
00000025 2016-01-21 19:39:11.320  3699  3699 "  /var/lib/HPCCSystems/myeclccserver/libW20160121-193910.so(+0x10b2) [0x7f1bfc2450b2]"
00000026 2016-01-21 19:39:11.320  3699  3699 "  /opt/HPCCSystems/lib/libworkunit.so(_ZN15WorkflowMachine11performItemEjj+0x7e) [0x7f1c038029fc]"
00000027 2016-01-21 19:39:11.320  3699  3699 "  /opt/HPCCSystems/lib/libworkunit.so(_ZN15WorkflowMachine13doExecuteItemER20IRuntimeWorkflowItemj+0x6a) [0x7f1c03801f80]"
00000028 2016-01-21 19:39:11.320  3699  3699 "  /opt/HPCCSystems/lib/libworkunit.so(_ZN15WorkflowMachine11executeItemEjj+0x20a) [0x7f1c0380199c]"
00000029 2016-01-21 19:39:11.320  3699  3699 "  /opt/HPCCSystems/lib/libworkunit.so(_ZN15WorkflowMachine7performEP18IGlobalCodeContextP11IEclProcess+0x1ad) [0x7f1c038014d5]"
0000002A 2016-01-21 19:39:11.320  3699  3699 "  /opt/HPCCSystems/lib/libhthor.so(_ZN8EclAgent10runProcessEP11IEclProcess+0x403) [0x7f1c05c75b1f]"
0000002B 2016-01-21 19:39:11.320  3699  3699 "  /opt/HPCCSystems/lib/libhthor.so(_ZN8EclAgent9doProcessEv+0x5f6) [0x7f1c05c748e4]"
0000002C 2016-01-21 19:39:11.320  3699  3699 "  /opt/HPCCSystems/lib/libhthor.so(_Z13eclagent_mainiPPKcP12StringBufferb+0x1824) [0x7f1c05c7dd50]"
0000002D 2016-01-21 19:39:11.320  3699  3699 "  eclagent(main+0x81) [0x401100]"
0000002E 2016-01-21 19:39:11.320  3699  3699 "  /lib/x86_64-linux-gnu/libc.so.6(__libc_start_main+0xf5) [0x7f1c04d50ec5]"
0000002F 2016-01-21 19:39:11.320  3699  3699 "  eclagent() [0x400fa9]"
00000030 2016-01-21 19:39:11.320  3699  3699 "ThreadList:
7F1BFF44D700 139758223546112 3700: CMPNotifyClosedThread
7F1BFEC4C700 139758215153408 3701: CSocketBaseThread
7F1BFE44B700 139758206760704 3702: MP Connection Thread
7F1BFDC4A700 139758198368000 3704: LogMsgParentReceiver
7F1BFD449700 139758189975296 3706: LogMsgFilterReceiver
7F1BFCC48700 139758181582592 3707: EclAgent Abort Monitor
7F1BF7FFF700 139758101591808 3708: CDaliPublisherClient
"
00000031 2016-01-21 19:39:11.321  3699  3699 "SIG: Aborted(6), accessing 000003E600000E73, IP=00007F1C04D65CC9"
```
So that is why I would like to add SIGABRT signal to our handler.